lib/virt_autotest/utils.pm

```diff
@@ -197,19 +198,21 @@ sub ensure_online {
     my $skip_network = $args{skip_network}  // 0;
     my $ping_delay   = $args{ping_delay}    // 15;
     my $ping_retry   = $args{ping_retry}    // 60;
+
     # Ensure guest is running
     # Only xen/kvm support to reboot guest at the moment
     if (is_xen_host || is_kvm_host) {
-        if (script_run("virsh list --all | grep '$guest' | grep running") != 0) {
+        if (script_run("virsh list | grep '$guest'") != 0) {
             assert_script_run("virsh start '$guest'");
+            script_retry("ping -c 1 '$guest'", delay => 10, retry => 30);
         }
     }
     unless ($skip_network == 1) {
         die "$guest does not respond to ICMP" if (script_retry("ping -c 1 '$guest'", delay => $ping_delay, retry => $ping_retry) != 0);
-        # Wait for ssh to come up
-        die "$guest does not start ssh" if (script_retry("nmap $guest -PN -p ssh | grep open", delay => 15, retry => 12) != 0);
         unless ($skip_ssh == 1) {
-            die "$guest not ssh-reachable" if (script_run("ssh $guest uname") != 0);
+            # Wait for ssh to come up
+            die "$guest does not start ssh" if (script_retry("nmap $guest -PN -p ssh | grep open", delay => 15, retry => 12) != 0);
```

Comment:
just an idea: is it a lit expensive to run nmap for this? alternatives `grep 'ssh' /etc/services` or `lsof -i :22`. wdyt?

Reply:
This is done for remote host. Before using actual `ssh` client we scan for the open port.